miden-tx/src/host/mod.rs

```diff
@@ -215,7 +216,14 @@ impl<A: AdviceProvider, T: TransactionAuthenticator> TransactionHost<A, T> {
         // update the delta tracker only if the current and new values are different
         if current_slot_value != new_slot_value {
             let slot_index = slot_index.as_int() as u8;
-            self.account_delta.storage_tracker().slot_update(slot_index, new_slot_value);
+            // if the new value is the new root of a StorageMap, there is no update of the
+            // delta tracker via slot_update. The map update will exclusively be tracked
+            // in maps_update. Any update with an index that is already in maps_update
+            // will be ignored. There should be never be a case where the map of a root is
+            // updated but there is no corresponding map update.
+            if !self.account_delta.storage_tracker().has_maps_update(&slot_index) {
+                self.account_delta.storage_tracker().slot_update(slot_index, new_slot_value);
+            }
```

Comment:
Thinking more about this, I wondering if this is sufficient. Specifically, what could happen is that the user could call `account::set_item` directly on a slot that contains a storage map and this won't result in an error (or at least not in an error here). This could just result in an error later or maybe it could corrupt the storage - but either way, I'm not sure that's good behavior.

Maybe a better approach is to modify the MASM after all. Specifically, maybe we should move toward the following:

- If `account::set_item` is called with an index which is not a simple value, the kernel should panic. This will probably require validating the slot type (which should already be in kernel memory).
- If `account::set_map_item` is called with an index which is not for a storage map, the kernel should panic. This will also require validating the slot type for the specified index.
- Both of the above procedures could call some internal procedure (e.g., `set_storage_item_raw`) which would update a given slot but would not perform any checks or fire any events.

We can also implement similar validation for `account::get_map_item`, but I think we can skip the validation for `account::get_item`.

If we do the above, the changes here may become unnecessary.

Reply:
We can do this for sure. However, I feel it is overly complicated, and the initial approach, including a check that the updated SlotIteam == new root, might be the better approach.

I changed it in the way you describe it here.